src/mrack/providers/beaker.py

```diff
@@ -0,0 +1,333 @@
+        beaker_id, req_hname = bkr_id_name
+
+        if not poll_sleep:
+            poll_sleep = self.poll_sleep
+        if not timeout:
+            timeout = self.timeout
+        if not max_attempts:
+            max_attempts = self.max_attempts
+
+        resource = {}
+        attempts = 0
+        prev_status = ""
+
+        while attempts < self.max_attempts:
+            attempts += 1
+            resource = self._get_recipe_info(beaker_id)
+            status = resource["status"]
+
+            if prev_status != status:
+                logger.info(
+                    f"Job {beaker_id} has changed status ("
+                    f"{prev_status} -> {status})"
+                )
+                prev_status = status
+
+            if STATUS_MAP.get(status) == STATUS_ACTIVE:
+                break
+            elif STATUS_MAP.get(status) == STATUS_PROVISIONING:
+                await asyncio.sleep(self.poll_sleep)
+            elif STATUS_MAP.get(status) in [STATUS_ERROR, STATUS_DELETED]:
+                logger.warning(
+                    f"Job {beaker_id} has errored with status "
+                    f"{status} with result {resource['result']}"
+                )
+                break
+            else:
+                logger.error(
+                    f"Job {beaker_id} has swithced to status "
+                    f"{status} with result {resource['result']}"
+                )
+                break
+
+        resource.update({"JobID": beaker_id, "hname": req_hname})
+        return resource
+
+    async def provision_hosts(self, hosts):
+        """Provision hosts based on list of host requirements.
+
+        Issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        started = datetime.now()
+
+        count = len(hosts)
+        logger.info(f"Issuing provisioning of {count} hosts")
+        create_res = []
+        for req in hosts:
+            resource = self.create_server(req)
+            create_res.append(resource)
+
+        create_resps = await asyncio.gather(*create_res)
+        logger.info("Provisioning issued")
+
+        logger.info("Waiting for all Beaker hosts to be available")
+        wait_res = []
+        for resp in create_resps:
+            resource = self.wait_till_provisioned(resp)
+            wait_res.append(resource)
+
+        server_results = await asyncio.gather(*wait_res)
+
+        provisioned = datetime.now()
+        provi_duration = provisioned - started
+
+        logger.info("All Beaker hosts reached provisioning final state (Reserved)")
+        logger.info(f"Provisioning duration: {provi_duration}")
+
+        hosts = [self.to_host(srv) for srv in server_results]
+        for host in hosts:
+            logger.info(host)
```

Comment:
Output of my test was:
```
Provisioning duration: 0:00:38.012023
error J:4536747 f30.aiohabit.test 10.XXX.XXX.6 None None
Error:
"Warn"
```
Last two lines came from Beaker client but they don't say a lot, did you see the same thing?

Reply:
Let us investigate this

Reply:
This was caused because the distro form metadata was not in prov-config and beaker selected default distro.
Beaker failed with: 
`Failed to provision recipeid 8753124, Failed to find repo for harness `
I will file an issue with that to have better handling for such cases.

Reply:
As discussed via IRC: This happened because I've requested a "wrong" distribution.

With a proper metadata.yaml file everything went well.
The error is still there for "wrong" scenarios. Perhaps this must be addressed by the `validate_hosts()` method, but we it can be done in a future PR.